orchagent/pfcwdorch.cpp

```diff
@@ -818,27 +879,89 @@ void PfcWdSwOrch<DropHandler, ForwardHandler>::doTask(swss::NotificationConsumer
 
             entry->second.handler->commitCounters();
             entry->second.handler = nullptr;
+            // Remove storm status in APPL_DB for warm-reboot purpose
+            string key = APP_PFC_WD_TABLE_NAME ":" + entry->second.portAlias;
+            m_applDbRedisClient.hdel(key, to_string(entry->second.index));
         }
     }
     else
     {
         SWSS_LOG_ERROR("Received unknown event from plugin, %s", event.c_str());
+        return false;
     }
+
+    return true;
 }
 
 template <typename DropHandler, typename ForwardHandler>
-void PfcWdSwOrch<DropHandler, ForwardHandler>::doTask(SelectableTimer &timer)
+bool PfcWdSwOrch<DropHandler, ForwardHandler>::bake()
 {
-    SWSS_LOG_ENTER();
+    // clean all *_last fields in COUNTERS_TABLE
+    // to allow warm-reboot pfc detect & restore logic to enter the same init state as cold-reboot
+    RedisClient redisClient(PfcWdOrch<DropHandler, ForwardHandler>::getCountersDb().get());
 
-    for (auto& handlerPair : m_entryMap)
+    vector<string> cKeys;
+    PfcWdSwOrch<DropHandler, ForwardHandler>::getCountersTable()->getKeys(cKeys);
+    for (const auto &key : cKeys)
     {
-        if (handlerPair.second.handler != nullptr)
+        vector<FieldValueTuple> fvTuples;
+        PfcWdSwOrch<DropHandler, ForwardHandler>::getCountersTable()->get(key, fvTuples);
+        for (const auto &fv : fvTuples)
         {
-            handlerPair.second.handler->commitCounters(true);
+            if (fvField(fv).find("_last") != string::npos)
+            {
+                redisClient.hdel(COUNTERS_TABLE ":" + key, fvField(fv));
+            }
         }
     }
 
+    // Re-organize the field-value to "storm" : <lossless queue list> (e.g., "3, 4")
+    Table table(m_applDb.get(), "_" APP_PFC_WD_TABLE_NAME);
```

Comment:
>"_" [](start = 32, length = 3)

Are you using the internal hash of ConsumerStateTable? It is not recommended.

Reply:
I transformed the syntax of field-value from queue_idx: storm to storm: queue_list, and used _PFC_WD_TABLE:Ethernet100 as a temp table to hold the new field-value (queue list in storm). Once refillToSync takes in the new field-value tuple, the temp table is removed. Not sure if it uses the internal hash of ConsumerStateTable

Reply:
Offline discussed, I understand you just create a temporary table and delete immediately later. I suggest you use original table schema, and in DoTask, you can find the table name.

Sample code: [https://github.com/Azure/sonic-swss/blob/a6d60f2ed0b0059be6014e2aba7b57f4dd7387a0/orchagent/portsorch.cpp#L2262](https://github.com/Azure/sonic-swss/blob/a6d60f2ed0b0059be6014e2aba7b57f4dd7387a0/orchagent/portsorch.cpp#L2262)

---
In reply to: [257486025](https://github.com/Azure/sonic-swss/pull/794#discussion_r257486025) [](ancestors = 257486025)

Reply:
The schema with the creation of a SubscriberStateTable is not supposed to hear anything anytime other than during the warm-reboot. This means that during the non-warm-reboot period, the storm status must not be updated to the subscribed table keyspace to avoid triggering the doTask, and in other words, this means that the storm status must be kept in a different keyspace, which is not being listened. Only when the warm-reboot is triggered, the subscribed keyspace will hear something, not natively but by explicitly refilled to it in an overrided bake(). While we save the transformation of the field-value  tuples in bake(), we still need to keep two keyspaces, one for warm-reboot doTask and one for tacking the current storm status, and the consumer object we created sits idle almost all the time only to fulfill its role for the flashing short-period of warm-reboot.